src/jit/lower.cpp

```diff
@@ -3761,6 +3843,197 @@ void Lowering::LowerAdd(GenTreePtr* pTree, Compiler::fgWalkData* data)
+            // computed once and stored in a temp variable.
+            CreateTemporary(&(divMod->gtOp.gtOp1));
+            dividend = divMod->gtGetOp1();
+
+            GenTreeStmt* curStmt = comp->compCurStmt->AsStmt();
+            unsigned curBBWeight = currBlock->getBBWeight(comp);
+            unsigned dividendLclNum = dividend->gtLclVar.gtLclNum;
+
+            GenTree* adjustment = comp->gtNewOperNode(
+                GT_RSH, type,
+                dividend,
+                comp->gtNewIconNode(type == TYP_INT ? 31 : 63));
+
+            if (absDivisorValue == 2)
+            {
+                // If the divisor is +/-2 then we'd end up with a bitwise and between 0/-1 and 1.
+                // We can get the same result by using GT_RSZ instead of GT_RSH.
+                adjustment->SetOper(GT_RSZ);
+            }
+            else
+            {
+                adjustment = comp->gtNewOperNode(
+                    GT_AND, type,
+                    adjustment,
+                    comp->gtNewIconNode(absDivisorValue - 1, type));
+            }
+
+            GenTree* adjustedDividend = comp->gtNewOperNode(
+                GT_ADD, type,
+                adjustment,
+                comp->gtNewLclvNode(dividendLclNum, type));
+
+            comp->lvaTable[dividendLclNum].incRefCnts(curBBWeight, comp);
+
+            GenTree* newDivMod;
+
+            if (isDiv)
+            {
+                // perform the division by right shifting the adjusted dividend
+                divisor->gtIntCon.SetIconValue(genLog2(absDivisorValue));
+
+                newDivMod = comp->gtNewOperNode(
+                    GT_RSH, type,
+                    adjustedDividend,
+                    divisor);
+
+                if (divisorValue < 0)
+                {
+                    // negate the result if the divisor is negative
+                    newDivMod = comp->gtNewOperNode(
+                        GT_NEG, type,
+                        newDivMod);
+                }
+            }
+            else
+            {
+                // divisor % dividend = dividend - divisor x (dividend / divisor)
+                // divisor x (dividend / divisor) translates to (dividend >> log2(divisor)) << log2(divisor)
+                // which simply discards the low log2(divisor) bits, that's just dividend & ~(divisor - 1)
+                divisor->gtIntCon.SetIconValue(~(absDivisorValue - 1));
+
+                newDivMod = comp->gtNewOperNode(
+                    GT_SUB, type,
+                    comp->gtNewLclvNode(dividendLclNum, type),
+                    comp->gtNewOperNode(
+                        GT_AND, type,
+                        adjustedDividend,
+                        divisor));
+
+                comp->lvaTable[dividendLclNum].incRefCnts(curBBWeight, comp);
+            }
+
+            // remove the divisor and dividend nodes from linear order so we can reuse them
```

Comment:
This comment might be clearer if it said `// Remove the divisor and dividend nodes from the linear order, since we have reused them and will resequence the tree` or something like that.